servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/ErrorClass.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+/**
+ * Enumeration of the main failure classes.
+ */
+enum ErrorClass {
+
+    /**
+     * Failures related to locally enforced timeouts that prevent session establishment with the peer.
+     */
+    LOCAL_ORIGIN_TIMEOUT(true),
+    /**
+     * Failures related to connection establishment.
+     */
+    LOCAL_ORIGIN_CONNECT_FAILED(true),
+
+    /**
+     * Failures related to locally enforced timeouts waiting for responses from the peer.
+     */
+    EXT_ORIGIN_TIMEOUT(false),
+
+    /**
+     * Failures returned from the remote peer. This will be things like 5xx responses.
+     */
+    EXT_ORIGIN_REQUEST_FAILED(false),
+
+    /**
+     * Failure due to cancellation.
+     */
+    CANCELLED(true),
+
+    /**
+     * Failure for an unknown reason.
+     */
+    UNKNOWN(false);
```

Comment:
There could be cases outside `LOCAL_ORIGIN_TIMEOUT` and `LOCAL_ORIGIN_CONNECT_FAILED` that are also local. For example, users may have a connection level filter that throws a runtime exception for all or some requests. Do you have an ideal how to distinguish local exceptions from those that caused by a remote peer?

Reply:
At the end of the day someone is going to have to classify exceptions into an error class. Ideally an exception would carry that info but right now it obviously doesn't.

Reply:
Because it doesn't (and likely won't be possible bcz users can inject any custom filter with any custom exception), will it make more sense to move `isLocal` knowledge outside `ErrorClass` constants into a method that does best effort to recognize the exception origin?

Reply:
In my minds eye the junction between the request/response pathway will  be in charge of converting the request/response/error pathway into the right `ErrorClass`. This can be done with a form of classifier that may be user configurable in the long term. The point of having the `isLocal` knowledge in the `ErrorClass` is so that it's not up to every implementation of `RequestTracker` to do that decoding themselves. We might even consider dropping the `Throwable` as a method param: I added it only for logging purposes but maybe it's not useful.